Fix things related to sort optimization

andreas-jonsson opened the pull request:
Some more fixes for yesterday's sort optimization PR.

Most commits should be self-explanatory.

6e84be2 fixes a bug where the sort would insert newline at the wrong place unless the entire file was sorted. To test, create a file like this:

```
1
2

3
```

Select the first two lines and sort in ascending fashion. Using yesterday's code, the file would end up like this (note that first line is an empty line):

```

1
2
3
```

But obviously 1-2 is already in ascending order and the file should have been left untouched.

andreas-jonsson commented:
eee7c4f is a new feature, cherry-pick it out if you don't want it.

Currently everything is sorted lexicographically, so an ascending sort of this:

```
90
89
9
```

is

```
89
9
90
```

I think most users would expect this result:

```
9
89
90
```

I have added logic to detect when all lines to be sorted are integers. If so, then we sort numerically and not lexicographically.

This could easily be extended to deal with decimal numbers as well (use std::stod instead of std::stoi), but it raises questions about how to deal with decimal point vs decimal comma, handling of spaces (i.e should we recognize "1 234,50" as a valid number), etc. Maybe a new option is needed for that, but there are already so many options in N++...

andreas-jonsson commented:
Note that I chose to let empty lines mean INT_MIN.

andreas-jonsson commented:
0eca4db is also related to numeric sort (I realized after committing that it could be improved).

andreas-jonsson commented:
Creating a vector of ints and then sorting is a lot faster than sorting a vector of strings directly and doing the conversion to int in the sorting lambda.

This requires that we be able to exactly roundtrip string -> int -> string. So we can't allow empty strings to count as INT_MIN (unless we add special logic). Not a big loss if you ask me.

I prepared a test file of 900 000 pseudorandom ints. On my computer I get these results:

Sorting before: > 10 seconds
Sorting after: 1.7 seconds

Could probably be sped up more (it does several passes through the lines), but I'm going to stop now and only deal with eventual bugs.

ariccio commented on PowerEditor/src/MISC/Common/Common.cpp:
```diff
@@ -747,4 +747,97 @@ generic_string stringJoin(const std::vector<generic_string> &strings, generic_st
+	else
+	{
+		// Check minus characters.
+		const int minuses = std::count(input.begin(), input.end(), TEXT('-'));
```
Neat use of `std::count`! I'll have to add that to my repertoire.

ariccio commented on PowerEditor/src/MISC/Common/Common.cpp:
```diff
@@ -747,4 +747,97 @@ generic_string stringJoin(const std::vector<generic_string> &strings, generic_st
+bool allLinesAreNumeric(const std::vector<generic_string> &lines)
+{
+	const auto endit = lines.end();
+	for (auto it = lines.begin(); it != endit; ++it)
```
Why the C++03 style iterator loop? There's nothing wrong with it, but I imagine a C++11 style ranged for would be cleaner here.

à la
```
for (const auto& line : lines)
{
    try
    {
        stoiStrict(line);
    }
    catch (std::invalid_argument&)
    {
        return false;
    }
    catch (std::out_of_range&)
    {
        return false;
    }
}
return true;
```

...totally nitpicking, of course.

(also)Might as well use `lines.cbegin()`/`lines.cend()`, if the ranged for doesn't seem better to you.

ariccio commented on PowerEditor/src/MISC/Common/Common.cpp:
```diff
@@ -747,4 +747,97 @@ generic_string stringJoin(const std::vector<generic_string> &strings, generic_st
+std::vector<generic_string> numericSort(std::vector<generic_string> input, bool isDescending)
+{
+	// Pre-condition: all strings in "input" are convertible to int with stoiStrict.
+	std::vector<int> inputAsInts;
```
How about a call to `inputAsInts.reserve(input.size())` here? Can't hurt.

ariccio commented on PowerEditor/src/MISC/Common/Common.cpp:
```diff
@@ -747,4 +747,97 @@ generic_string stringJoin(const std::vector<generic_string> &strings, generic_st
+			return a < b;
+		}
+	});
+	std::vector<generic_string> output;
```
(same as above, about `reserve`).

ariccio commented on PowerEditor/src/MISC/Common/Common.h:
```diff
@@ -188,7 +188,12 @@ generic_string PathAppend(generic_string &strDest, const generic_string & str2ap
-std::vector<generic_string> stringSplit(const generic_string& input, generic_string delimiter);
-generic_string stringJoin(const std::vector<generic_string> &strings, generic_string separator);
+std::vector<generic_string> stringSplit(const generic_string& input, const generic_string &delimiter);
+generic_string stringJoin(const std::vector<generic_string> &strings, const generic_string &separator);
```
`const std::vector<generic_string>& input` or `const std::vector<generic_string> &input`?

Consistency is good.

ariccio commented on PowerEditor/src/ScitillaComponent/ScintillaEditView.cpp:
```diff
@@ -1922,6 +1922,17 @@ void ScintillaEditView::getLine(int lineNumber, TCHAR * line, int lineBufferLen)
+{
+	int lineLen = execute(SCI_LINELENGTH, lineNumber);
+	const int bufSize = 1 + lineLen;
+	_TCHAR *buf = new _TCHAR[bufSize];
```
You changed `_TCHAR` to `TCHAR` above, but used `_TCHAR` here?

Huh?

ariccio commented on PowerEditor/src/ScitillaComponent/ScintillaEditView.cpp:
```diff
@@ -2958,19 +2969,35 @@ void ScintillaEditView::quickSortLines(size_t fromLine, size_t toLine, bool isDe
 	std::vector<generic_string> splitText = stringSplit(text, getEOLString());
-	std::sort(splitText.begin(), splitText.end(), [isDescending](generic_string a, generic_string b)
+	const size_t lineCount = execute(SCI_GETLINECOUNT);
+	const bool sortEntireDocument = toLine == lineCount - 1;
```
I'm liking the `const` :)

ariccio commented:
> I prepared a test file of 900 000 pseudorandom ints. 

(one of) the mark(s) of a good computer scientists is supporting assertions/hypothesis with evidence. Nice work.

Lgtm. +1.

ariccio commented:
I'm curious though, I didn't see any exception safety problems here (hence the *lgtm*), but why the manual memory management?

andreas-jonsson commented:
Have fixed these things now.

The only manual memory management I can see is in getLine(). I don't see how it can be written without new/delete since it interfaces with Scintilla. But it's moot anyway since I have removed the function since it wasn't used.

------

I have also fixed numeric sort so it can handle empty lines again.

ariccio commented:
> Have fixed these things now.

Nice work. Good patch! 

I can't get over that use of `std::count` - it's awesome.

-----
> I don't see how it can be written without new/delete since it interfaces with Scintilla.

Eh? I don't see a problem with `std::unique_ptr` here? It is moot indeed, but wouldn't it be:

```
	std::unique_ptr<TCHAR[]> buf = std::make_unique<TCHAR[]>(bufSize);
 	getGenericText(buf.get( ), bufSize, start, end);
 	generic_string text = buf.get( );
```

andreas-jonsson commented:
You're right, I thought you meant that it was somehow possible to get a std::wstring directly out of Scintilla.
